### PR TITLE
Added Synthetic Features

### PR DESCRIPTION
**Pull request by francocm**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -243,3 +243,5 @@ TSWLatexianTemp*
 
 # generated if using elsarticle.cls
 *.spl
+
+.DS_Store
```
Comment (jp-um) [inline .gitignore]: Good idea ... but you should do this on an Ubuntu machine ... not a Mac ...

File: terms/syntheticfeatures_fcm.bib [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  booktitle = {Proceedings of the ACM SIGKDD International Conference on Knowledge Discovery and Data Mining},
+  doi = {10.1145/1401890.1401959},
+  isbn = {9781605581934},
+  issn = {0309-0167 (Print)},
+  keywords = {,classification rules,discrimination},
+  number = {May 2014},
+  pages = {560--568},
+  pmid = {7698734},
+  title = {{Discrimination-aware data mining}},
+  url = {http://dl.acm.org/citation.cfm?doid=1401890.1401959},
```
Comment (jp-um) [inline terms/syntheticfeatures_fcm.bib]: Did you read the instructions in the README file ???

> .bib references do not need DOI, keywords or abstract entries (other attributes to delete are Remove Date-Added, Date-Modified, Bdsk-File-1)

Also remove url.  Not needed.

File: terms/syntheticfeatures_fcm.tex [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+To mitigate this problem, a possible solution is to generate and use synthetic features. This gives the researcher the freedom to generate data that fits the scope of the problem at hand, without access to the original data, whilst still allowing for proper evaluation of different techniques and algorithms \citep{patki2016}.
+
+Over and above to unavailability of data, such an approach can also be used to enhance and augment limited data that might be available, such as synthetic over-sampling in case of unbalanced datasets \citep{tallo2018}.
+
+Such an approach presents itself with a number of advantages for evaluation models and algorithms:
+
+\begin{itemize}
+    \item Evaluation of the model’s behaviour and performance under specific data conditions and complexities, and not just over the standard statistical distribution of a real dataset.
```
Comment (jp-um) [inline terms/syntheticfeatures_fcm.tex]: Have you read this to check its fine?  Look at what is generated for `model’s`.    You'll see this is broken in the pdf.  Fix.

File: terms/syntheticfeatures_fcm.tex [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    \end{itemize}
+    \item Allow the researcher to focus on the actual problem at hand (the ML problem) rather than how to obtain access to the training data.
+    \begin{itemize}
+        \item When evaluating models, you would typically first want to determine the best approach for your problem, rather than focusing on training and delivering the final model. That comes later, after these technical issues are resolved.
+    \end{itemize}
+\end{itemize}
+
+\section{Possible Solutions}
+
+There are various types of synthetic data that could be generated, and typically the researcher would need to take a decision on which to make use of, or possibly even a combination of them, to target the kind of problem they’re trying to solve.
```
Comment (jp-um) [inline terms/syntheticfeatures_fcm.tex]: Again, have you read the guidelines?

> No abbreviations (e.g. they're), but also no Fig., Eq. etc. This is a formal scientific document, not an email.

File: terms/syntheticfeatures_fcm.tex [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+For this reason, a number of toolkits exist that offer a more focused or targeted generation mechanism for datasets, which allow for greater control and tuning as to how data is outputted and shaped. Typically, this means that the researcher has the ability to define the problem types as well as specific parameters (boundaries, noise, direction, classes, etc) for each type. These include:
+
+\begin{itemize}
+    \item Regression data
+    \item Classification data
+    \item Clustering data
+\end{itemize}
+
+For each of the above, the generated datasets can be “perfect”, or an element of noise and outliers included as well to simulate a more real-life dataset.
```
Comment (jp-um) [inline terms/syntheticfeatures_fcm.tex]: Look at how those double quotes look in the pdf (BROKEN!)

File: .gitignore (hunk 1 above)
Comment (francocm) [reply .gitignore]: Hehe yes. Main PC (this I'm writing this comment on) is an Ubuntu. Mac is the side laptop. Don't worry I'm a Linux guy ;)

**Comment by francocm** [comment]
Ah. I had understood synthetic features as described in the context of what I wrote so far (dummy data generation), and not as a feature engineering point of view in the sense of extrapolating / deriving features from existing data (feature engineering). Am familiar with this topic too. Will update it over the weekend. Thanks!

**Comment by francocm** [comment]
@jp-um Updated for reconsideration. Thanks.

**Comment by jp-um** [comment]
You have supplied a great example (well done), but you fail to go in the more theoretical aspects ... what are the ways in which you can generate synthetic features.  Do you want to update your entry?

Also fix the first paragraph -- it is not maturely written (and you like the word essentially too much).

**Comment by francocm** [comment]
@jp-um Thanks. Updated accordingly. Could you kindly re-evaluate the pull request please? Thanks

File: terms/syntheticfeatures_fcm.bib [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  isbn = {978-1-59593-597-7},
+  pages = {871--872},
+  publisher = {ACM},
+  series = {SIGIR '07},
+  title = {{Feature Engineering for Mobile (SMS) Spam Filtering}},
+  url = {http://doi.acm.org/10.1145/1277741.1277951},
+  year = {2007}
+}
+
+@inproceedings{LiFeatureEng,
```
Comment (jp-um) [inline terms/syntheticfeatures_fcm.bib]: url is not needed for article and inproceedings entries ...  remove

File: terms/syntheticfeatures_fcm.tex [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+      \textit & \makecell[tl]{Content} & \makecell[tl]{The comment body definitely has a big relevance in the classification result, \\ however, can the whole sentence be easily understood by the algorithm as \\ it is?} \\
+      \bottomrule
+    \end{tabular}}
+    \label{tab:sf_origdataexplained}
+\end{table}
+\vspace{2mm}
+
+As one can see, there is very little input the machine learning algorithm can reliably take just from using the four features described above. One could easily realise this by asking oneself the following question (in plain English):
+
+\qquad \emph{How can I describe the components of a comment well enough to decide whether it is probably spam or ham?}
```
Comment (jp-um) [inline terms/syntheticfeatures_fcm.tex]: remove manual spacing

File: terms/syntheticfeatures_fcm.tex [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+What or how an explosion of features or a composite of features is generated depends on the very specific nature of the components involved and there is no generic formula behind it that works without some additional specificness. For example, two pairs of geo-coordinates probably qualify in giving a distance feature, however the formula applied here is specific to the geo-coordinates domain.
+
+There is not a one-size-fits-all approach but rather it is more of an iterative approach with new synthetic features being outputted per iteration, following which one then assesses whether it is enough to generate a reliable machine learning model from the new features or not.
+
+Following below is a practical example of this technique, using the dataset described at the introduction of this chapter.
+
+\subsection{Analysing each feature} \index{synthetic-features!solving-problem-analysing-each-feature}
+
+\textbf{Video}
```
Comment (jp-um) [inline terms/syntheticfeatures_fcm.tex]: what is this?  use \subsubsection dont invent your own titling.  Also remove all manual spacing (`\\`) see guideline 31 in README.

File: terms/syntheticfeatures_fcm.tex [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+\subsection{Updated feature / data set} \index{synthetic-features!updated-feature-dataset}
+
+Following the synthetic feature generation described above, the updated data set used as an example here would look as follows:
+
+\begin{table}[ht]
+    \centering
+    \fontfamily{ppl}\selectfont
+    \resizebox{\textwidth}{!}{\begin{tabular}{llllllllll}
+      \toprule
+                        & \textit{\makecell{Video}} & \textit{\makecell{Author Comments \\ in last minute}} & \textit{\makecell{Language}} & \textit{\makecell{Readability}} & \textit{\makecell{Length excl. \\ stop words}} & \textit{\makecell{Presence of \\ account tags}} & \textit{\makecell{Presence of \\ URLs}} & \textit{\makecell{Presence of \\ emojis}} & \textit{\makecell{Class}} \\
```
Comment (jp-um) [inline terms/syntheticfeatures_fcm.tex]: textit should be with each text not on the whole row.  look at examples

File: terms/syntheticfeatures_fcm.tex (hunk 8 above)
Comment (francocm) [reply terms/syntheticfeatures_fcm.tex]: `\subsubsection` is not supported by Tufte as its philosophy discourages such subsections. I attempted to use it anyway but it failed with the error:

```
Class tufte-book Error: 
(tufte-book)              \subsubsectionis undefined by this class. See Robert Bringhurst's _The Elements of Typographic Style_, section 4.2.2. \subsubsectionwas used
```

See:
* [Tufte issue 44](https://github.com/rstudio/tufte/issues/44)
* [Tufte issue 34](https://github.com/rstudio/tufte/issues/34)

On this note, I'm not sure who I can proceed on it and leaving it as is for now unless you can suggest otherwise.

**Comment by francocm** [comment]
Changes done as highlighted plus added a question above regarding `\subsubsection`.

Moreover there was an issue with a broken unicode character inside [terms/regularisationofmodels_ja.tex line 60](https://github.com/jp-um/machine_learning_handbook/pull/24/commits/b912dd4753a8bdc1e444165c506aaff568455330#diff-5a395e1d5765d802fe572c1a6e3e02a6R60)
where I changed `overﬁtting` to `overfitting` (single fi character) as it was breaking the build.

File: terms/syntheticfeatures_fcm.tex (hunk 8 above)
Comment (francocm) [reply terms/syntheticfeatures_fcm.tex]: Fixed by converting to lists.

**Comment by francocm** [comment]
Issue with `\subsubsection` resolved by using lists.

**Comment by jp-um** [comment]
Thanks, closed.
